x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/assignment/planning/RandomizedAssignmentRounding.java

```diff
@@ -0,0 +1,428 @@
+        }
+
+        private AssignmentHolder(AssignmentHolder holder) {
+            softAssignments.putAll(holder.softAssignments);
+            softAllocations.putAll(holder.softAllocations);
+            resourceTracker = new ResourceTracker(holder.resourceTracker);
+        }
+
+        private void initializeAssignments(Map<Tuple<Model, Node>, Double> allocationVars, Map<Tuple<Model, Node>, Double> assignmentVars) {
+            for (Node n : nodes) {
+                for (Model m : models) {
+                    Tuple<Model, Node> index = Tuple.tuple(m, n);
+                    double assignment = assignmentVars.get(index);
+                    double allocations = allocationVars.get(index);
+
+                    if (assignment == 1.0 && isInteger(allocations)) {
+                        resourceTracker.assign(m, n, (int) Math.rint(allocations));
+                    }
+                    softAssignments.put(index, assignment);
+                    softAllocations.put(index, allocations);
+                }
+            }
+        }
+
+        private void assignUnderSubscribedNodes() {
+            assignUnderSubscribedNodes(nodes);
+        }
+
+        private void assignUnderSubscribedNodes(Collection<Node> nodeSelection) {
+            // Snap to one any non-zero assignments on nodes where all the soft assigned models fit.
+            for (Node n : nodeSelection.stream().sorted(Comparator.comparingDouble(this::decreasingQualityNodeOrder)).toList()) {
+                List<Model> assignedModels = new ArrayList<>();
+                long totalModelMemory = 0;
+                int maxTotalThreads = 0;
+                for (Model m : models) {
+                    Tuple<Model, Node> assignment = Tuple.tuple(m, n);
+                    if (softAssignments.get(assignment) > 0) {
+                        totalModelMemory += m.memoryBytes();
+                        maxTotalThreads += (int) Math.ceil(softAllocations.get(assignment)) * m.threadsPerAllocation();
+                        assignedModels.add(m);
+                    }
+                }
+                if (totalModelMemory <= n.availableMemoryBytes() && maxTotalThreads <= n.cores()) {
+                    for (Model m : assignedModels) {
+                        Tuple<Model, Node> index = Tuple.tuple(m, n);
+
+                        int allocationsToAssign = 0;
+                        if (isInteger(softAllocations.get(index))) {
+                            // We round this separately because if we used ceil and the value was just about the
+                            // integer value we'll use one additional allocation when we shouldn't.
+                            allocationsToAssign = (int) Math.rint(softAllocations.get(index));
+                        } else if (maxTotalThreads <= n.cores()) {
+                            allocationsToAssign = (int) Math.ceil(softAllocations.get(index));
+                        }
```

Comment:
I think we should use the same logic to compute `maxTotalThreads` as we use here, i.e. with rounding. Also `maxTotalThreads <= n.cores()` is redundant since you only enter if this is true. I would factor out into a method:
```
int allocationsToAssign(...) {
    if (isInteger(softAllocations.get(index))) {
        // We round this separately because if we used ceil and the value was just about the
        // integer value we'll use one additional allocation when we shouldn't.
        return (int) Math.rint(softAllocations.get(index));
    }
    return (int) Math.ceil(softAllocations.get(index));
}
```